src/functional_tests.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::{
+    actions::*,
+    state::*,
+    utilities::{
+        Candidate, Node, NodeState, ParsecVote, RelocatedInfo, Rpc, State, TestEvent, TryResult,
+    },
+};
+use rand::{self, Rng, SeedableRng};
+use rand_xorshift::XorShiftRng;
+use std::{env, iter, thread};
+use unwrap::unwrap;
+
+fn get_rng() -> XorShiftRng {
+    let env_var_name = "ROUTING_MODEL_SEED";
+    let seed = env::var(env_var_name)
+        .ok()
+        .map(|value| {
+            unwrap!(
+                value.parse::<u64>(),
+                "Env var 'ROUTING_MODEL_SEED={}' is not a valid u64.",
+                value
+            )
+        })
+        .unwrap_or_else(rand::random);
+    println!(
+        "To replay this '{}', set env var {}={}",
+        unwrap!(thread::current().name()),
+        env_var_name,
+        seed
+    );
+    XorShiftRng::seed_from_u64(seed)
+}
+
+#[test]
+fn relocate_adult_src() {
+    let adult_count = 2;
+    let elder_count = 3;
+    let mut rng = get_rng();
+    let mut nodes = iter::repeat_with(|| rng.gen())
+        .take(adult_count + elder_count)
+        .collect::<Vec<Node>>();
+    nodes.sort_unstable();
+    let (adults, elders) = nodes.split_at(adult_count);
+
+    let action = Action::new(
+        InnerAction::new_with_our_attributes(rng.gen())
+            .with_next_target_interval(rng.gen())
+            .extend_current_nodes_with(&NodeState::default_elder(), elders)
+            .extend_current_nodes_with(&NodeState::default(), adults),
+    );
+    let mut member_state = MemberState {
+        action,
+        ..Default::default()
+    };
+
+    let relocating_node = adults[0];
+    assert!(member_state
+        .action
+        .inner()
+        .our_current_nodes
+        .contains_key(&relocating_node.0.name));
+
+    let relocated_info = RelocatedInfo {
+        candidate: Candidate(relocating_node.0),
+        expected_age: relocating_node.0.age.increment_by_one(),
+        target_interval_centre: rng.gen(),
+        section_info: rng.gen(),
+    };
+
+    let required_events = [
+        TestEvent::SetWorkUnitEnoughToRelocate(relocating_node).to_event(),
+        ParsecVote::WorkUnitIncrement.to_event(),
+        ParsecVote::CheckRelocate.to_event(),
+        Rpc::RelocateResponse(relocated_info).to_event(),
```

Comment:
This is not a Requiered event I think

Reply:
Hmm, yes, that's true.  I guess I was thinking of events which could validly be triggered while we're in this flow (I've called them `optional_random_events`) and the others I just called `required` since they're not optional for the purposes of this test.

`Rpc::RelocateResponse` falls somewhere in the middle I suppose.  The test would probably still pass if we just moved that to the optional events, but in real life it'll never get handled more than once.

Not sure if we should be that specific here though.  If the flow can handle duplicate incoming section RPCs, and the tests can too, then maybe it's ok to just add it to the optional ones?

Reply:
The flow allow to send it more than once as we may have send multiple requests, and the response could always be replayed.

Reply:
> and the response could always be replayed

That's the part that Routing currently shields us from.  But allowing for it in the flows means we should test for it here I suppose.

Reply:
My understanding it routing does a best effort (after time out replay can occur).

Reply:
Yes, but the timeout should be large enough that it's infeasible a replay occurs.